pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,738 @@
+				}
+			}
+			for idx := range dnsrec {
+				for _, dnsRec := range dnsrec[idx] {
+					// temp print out all unique dns replies for the bridge
+					log.Infof("!!FlowStats: DNS time %v, domain %s, appIP %v, count %d, Answers %v",
+						dnsRec.TimeStamp, dnsRec.DomainName, dnsRec.AppIP, dnsRec.ANCount, dnsRec.Answers)
+
+					dnsrec := types.DNSReq{
+						HostName:    dnsRec.DomainName,
+						Addrs:       dnsRec.Answers,
+						RequestTime: dnsRec.TimeStamp.UnixNano(),
+					}
+					dnsPacked = true
+					flowdata.DNSReqs = append(flowdata.DNSReqs, dnsrec)
+				}
+			}
+
+			// flow record done for the bridge/app
+			// publish the flow data (per app/bridge) to zedagent now
+			flowKey := scope.UUID.String() + scope.NetUUID.String()
+			pub.Publish(flowKey, &flowdata)
+		}
+	}
+
+	// remove the dns data already uploaded
+	if dnsPacked {
+		for bnx := range instData.bnNet {
+			bnNum, err := bridgeStrToNum(bnx)
+			if err != nil {
+				continue
+			}
+			dnssys[bnNum].Snoop = nil
+			log.Infof("!!FlowStats: clear dns record for bn%d", bnNum)
+		}
+	}
+}
+
+// conntrack flow of two uni-directional stats into one
+// bireditional flow stats
+func flowMergeProcess(entry *netlink.ConntrackFlow, instData networkAttrs) flowStats {
+	var ipFlow flowStats
+	var forwSrcApp, forwDstApp, backSrcApp, backDstApp bool
+	var AppNum int
+
+	// for the current phase of implementation, we ignore the flowstats which has not timed out yet
+	ipflowTimeOut := int32(entry.TimeOut)
+	if ipflowTimeOut > timeoutSec {
+		return ipFlow
+	}
+
+	ipFlow.aclNum = entry.Mark & markMask
+	ipFlow.appNum = uint8(entry.Mark >> appShiftBits)
+	AppNum = int(ipFlow.appNum)
+	if AppNum == 0 { // only handle App related flow stats
+		return ipFlow
+	}
+
+	ipFlow.TimeStart = int64(entry.TimeStart)
+	ipFlow.TimeStop = int64(entry.TimeStop)
+	ipFlow.TimeOut = entry.TimeOut
+	ipFlow.Proto = entry.Forward.Protocol
+	ipFlow.IsTimeOut = true
+
+	// Assume the App has an assigned IP address(es) first
+	// the instData.appIPAddrs has the IP addresses of an App, we want to know
+	// which one of the 4 IP addresses of the flow tuple matches the App IPs
+	if len(instData.appIPAddrs[AppNum]) > 0 {
+		ipFlow.dbg1 = 1
+		forwSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.SrcIP)
+		if forwSrcApp == false {
+			ipFlow.dbg1 = 2
+			forwDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Forward.DstIP)
+			if forwDstApp == false {
+				ipFlow.dbg1 = 3
+				backSrcApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.SrcIP)
+				if backSrcApp == false {
+					ipFlow.dbg1 = 4
+					backDstApp = checkAppIPAddr(instData.appIPAddrs[AppNum], entry.Reverse.DstIP)
+				} else {
+					ipFlow.dbg1 = 5 // XXX
+				}
+			}
+		}
+	}
+
+	// if failed to get an App IP to match flow tuple, find remote from Intf subnets
+	// and reversely getting the 'app' IP
+	// Find which endpoint of the flow is NOT on my IP subnets, assume that is the
```

Comment:
How does this work for local traffic between two app instances where isInMySubnets would match for both source and dest? 
Do we want to have a different dbg1 value for that case?

Reply:
the current code goes through the list, the first match will jump out, then declare this is the App. From merge two unidirectional flow point of view it works if the other end is also an App. From debug point of view, right we don't catch this case.